File: ios-app/LaVeinteDigital/IMSS/Tarjeton/TarjetonDigitalDelegaciones.swift
import Foundation

/// Catálogo de delegaciones de Tarjetón Digital IMSS (port de
/// `TarjetonDigitalDelegaciones.kt`). El listado estático es respaldo offline;
/// el portal manda (se refresca desde `<select id="ddlDelegacion">`).
enum TarjetonDigitalDelegaciones {

    struct Delegacion: Equatable {
        let value: String
        let label: String
        var displayName: String { TarjetonDigitalDelegaciones.prettify(label) }
    }

    static let fallback: [Delegacion] = [
        Delegacion(value: "01", label: "AGUASCALIENTES"),
        Delegacion(value: "02", label: "BAJA CALIFORNIA"),
        Delegacion(value: "03", label: "BAJA CALIFORNIA SUR"),
        Delegacion(value: "04", label: "CAMPECHE"),
        Delegacion(value: "05", label: "COAHUILA"),
        Delegacion(value: "06", label: "COLIMA"),
        Delegacion(value: "07", label: "CHIAPAS"),
        Delegacion(value: "08", label: "CHIHUAHUA"),
        Delegacion(value: "09", label: "OFICINAS CENTRALES"),
        Delegacion(value: "10", label: "DURANGO"),
        Delegacion(value: "11", label: "GUANAJUATO"),
        Delegacion(value: "12", label: "GUERRERO"),
        Delegacion(value: "13", label: "HIDALGO"),
        Delegacion(value: "14", label: "JALISCO"),
        Delegacion(value: "15", label: "ESTADO DE MEXICO ORIENTE"),
        Delegacion(value: "16", label: "ESTADO DE MEXICO PONIENTE"),
        Delegacion(value: "17", label: "MICHOACAN"),
        Delegacion(value: "18", label: "MORELOS"),
        Delegacion(value: "19", label: "NAYARIT"),
        Delegacion(value: "20", label: "NUEVO LEON"),
        Delegacion(value: "21", label: "OAXACA"),
        Delegacion(value: "22", label: "PUEBLA"),
        Delegacion(value: "23", label: "QUERETARO"),
        Delegacion(value: "24", label: "QUINTANA ROO"),
        Delegacion(value: "25", label: "SAN LUIS POTOSI"),
        Delegacion(value: "26", label: "SINALOA"),
        Delegacion(value: "27", label: "SONORA"),
        Delegacion(value: "28", label: "TABASCO"),
        Delegacion(value: "29", label: "TAMAULIPAS"),
        Delegacion(value: "30", label: "TLAXCALA"),
        Delegacion(value: "31", label: "VERACRUZ NORTE"),
        Delegacion(value: "32", label: "VERACRUZ SUR"),
        Delegacion(value: "33", label: "YUCATAN"),
        Delegacion(value: "34", label: "ZACATECAS"),
        Delegacion(value: "35", label: "35 NORTE DEL DISTRITO FEDERAL"),
        Delegacion(value: "36", label: "36 NORTE DEL DISTRITO FEDERAL"),
        Delegacion(value: "37", label: "37 SUR DEL DISTRITO FEDERAL"),
        Delegacion(value: "38", label: "38 SUR DEL DISTRITO FEDERAL"),
    ]

    private static let particles: Set<String> = ["de", "del", "la", "el", "y", "los", "las"]

    /// Convierte la etiqueta en mayúsculas del portal a un nombre amigable con
    /// acentos correctos (MICHOACAN → Michoacán, NUEVO LEON → Nuevo León, etc.).
    static func prettify(_ raw: String) -> String {
        let normalized = raw
            .trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .lowercased()
        let accented = knownAccents[normalized] ?? normalized
        let words = accented.split(separator: " ")
        return words.enumerated().map { (i, word) in
            let w = String(word)
            if w.isEmpty { return w }
            if i > 0 && particles.contains(w) { return w }
            return w.prefix(1).uppercased() + w.dropFirst()
        }.joined(separator: " ")
    }

    private static let knownAccents: [String: String] = [
        "michoacan": "michoacán",
        "nuevo leon": "nuevo león",
        "queretaro": "querétaro",
        "yucatan": "yucatán",
        "san luis potosi": "san luis potosí",
        "estado de mexico oriente": "estado de méxico oriente",
        "estado de mexico poniente": "estado de méxico poniente",
    ]
}

File: android-app/app/src/main/java/com/laveintedigital/app/imss/tarjeton/TarjetonDigitalDelegaciones.kt
package com.laveintedigital.app.imss.tarjeton

/**
 * Catálogo de delegaciones de Tarjetón Digital IMSS.
 *
 * Los valores (`value` / `label`) provienen del portal real: la lista la entrega
 * `wsRegistroUsuario.asmx/wsConsultaDelegaciones` y la APK la refresca en vivo desde
 * el `<select id="ddlDelegacion">` cuando el WebView carga el formulario. Este listado
 * estático es únicamente un respaldo offline para que el formulario nativo muestre
 * nombres amigables desde el primer frame.
 *
 * `value` = el `Valor` real del portal (texto del atributo value del `<option>`).
 * `label` = el `Texto` real del portal (en mayúsculas, sin acentos).
 */
object TarjetonDigitalDelegaciones {

    data class Delegacion(val value: String, val label: String) {
        /** Nombre amigable para la interfaz: "Estado de México Oriente", "Michoacán". */
        val displayName: String get() = prettify(label)
    }

    /** Respaldo offline — NO es autoritativo; el portal manda. */
    val FALLBACK: List<Delegacion> = listOf(
        Delegacion("01", "AGUASCALIENTES"),
        Delegacion("02", "BAJA CALIFORNIA"),
        Delegacion("03", "BAJA CALIFORNIA SUR"),
        Delegacion("04", "CAMPECHE"),
        Delegacion("05", "COAHUILA"),
        Delegacion("06", "COLIMA"),
        Delegacion("07", "CHIAPAS"),
        Delegacion("08", "CHIHUAHUA"),
        Delegacion("09", "OFICINAS CENTRALES"),
        Delegacion("10", "DURANGO"),
        Delegacion("11", "GUANAJUATO"),
        Delegacion("12", "GUERRERO"),
        Delegacion("13", "HIDALGO"),
        Delegacion("14", "JALISCO"),
        Delegacion("15", "ESTADO DE MEXICO ORIENTE"),
        Delegacion("16", "ESTADO DE MEXICO PONIENTE"),
        Delegacion("17", "MICHOACAN"),
        Delegacion("18", "MORELOS"),
        Delegacion("19", "NAYARIT"),
        Delegacion("20", "NUEVO LEON"),
        Delegacion("21", "OAXACA"),
        Delegacion("22", "PUEBLA"),
        Delegacion("23", "QUERETARO"),
        Delegacion("24", "QUINTANA ROO"),
        Delegacion("25", "SAN LUIS POTOSI"),
        Delegacion("26", "SINALOA"),
        Delegacion("27", "SONORA"),
        Delegacion("28", "TABASCO"),
        Delegacion("29", "TAMAULIPAS"),
        Delegacion("30", "TLAXCALA"),
        Delegacion("31", "VERACRUZ NORTE"),
        Delegacion("32", "VERACRUZ SUR"),
        Delegacion("33", "YUCATAN"),
        Delegacion("34", "ZACATECAS"),
        Delegacion("35", "35 NORTE DEL DISTRITO FEDERAL"),
        Delegacion("36", "36 NORTE DEL DISTRITO FEDERAL"),
        Delegacion("37", "37 SUR DEL DISTRITO FEDERAL"),
        Delegacion("38", "38 SUR DEL DISTRITO FEDERAL"),
    )

    /** Partículas que van en minúscula dentro del nombre (título en español). */
    private val particles = setOf("de", "del", "la", "el", "y", "los", "las")

    /**
     * Convierte la etiqueta en mayúsculas del portal a un nombre amigable con
     * acentos correctos en español. Maneja casos conocidos donde el portal omite
     * tildes (MICHOACAN → Michoacán, NUEVO LEON → Nuevo León, etc.).
     */
    fun prettify(raw: String): String {
        val normalized = raw.trim().replace(Regex("""\s+"""), " ").lowercase()
        val accented = knownAccents[normalized] ?: normalized
        val words = accented.split(" ")
        return words.mapIndexed { i, word ->
            if (word.isEmpty()) word
            else {
                val cased = word.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
                // Primera palabra siempre con mayúscula; las partículas intermedias en minúscula.
                if (i > 0 && word in particles) word else cased
            }
        }.joinToString(" ")
    }

    /** Acentos conocidos que el portal escribe sin tilde (clave en minúsculas). */
    private val knownAccents: Map<String, String> = mapOf(
        "michoacan" to "michoacán",
        "nuevo leon" to "nuevo león",
        "queretaro" to "querétaro",
        "yucatan" to "yucatán",
        "san luis potosi" to "san luis potosí",
        "estado de mexico oriente" to "estado de méxico oriente",
        "estado de mexico poniente" to "estado de méxico poniente",
    )
}
